### PR TITLE
Add support for pulser-diff backend

### PR DESCRIPTION
**Pull request by vytautas-a**
Modify `pulser-core` to support all current features of differentiable `torch` backend.

### CODE REVIEW

File: pulser-core/pulser/channels/dmm.py [hunk 1]
```diff
@@ -124,7 +125,9 @@ def validate_pulse(
                 (defaults to a detuning map with weight 1.0).
         """
         super().validate_pulse(pulse)
-        round_detuning = round(pulse.detuning.samples, 6).as_array(detach=True)
+        round_detuning = pm.round(pulse.detuning.samples, 6).as_array(
+            detach=True
+        )
```
Comment (HGSilveri) [inline pulser-core/pulser/channels/dmm.py:130]: What was the issue with the previous version?

File: pulser-core/pulser/math/abstract_array.py [hunk 2]
```diff
@@ -77,7 +77,7 @@ def as_tensor(self) -> torch.Tensor:
             raise RuntimeError("`torch` is not installed.")
         return torch.as_tensor(self._array)
 
-    def as_array(self, detach: bool = False) -> np.ndarray:
+    def as_array(self, detach: bool = True) -> np.ndarray:
```
Comment (HGSilveri) [inline pulser-core/pulser/math/abstract_array.py:80]: I thought we decided that we wouldn't detach by default

File: pulser-core/pulser/pulse.py [hunk 3]
```diff
@@ -103,10 +103,7 @@ def __init__(
             )
         object.__setattr__(self, "amplitude", amplitude)
         object.__setattr__(self, "detuning", detuning)
-        assert not isinstance(phase, Parametrized)
-        if (phase_ := pm.AbstractArray(phase, dtype=float)).ndim != 0:
-            raise ValueError(f"'phase' must be a single float, not {phase!r}.")
-        object.__setattr__(self, "phase", phase_ % (2 * np.pi))
+        object.__setattr__(self, "phase", cast(float, phase) % (2 * np.pi))
```
Comment (HGSilveri) [inline pulser-core/pulser/pulse.py:106]: I think what I had before is what we want here

File: pulser-core/pulser/waveforms.py [hunk 4]
```diff
@@ -509,7 +509,9 @@ def _samples(self) -> pm.AbstractArray:
         Returns:
             A numpy array with a value for each time step.
         """
-        return self._value * np.ones(self.duration)
+        res = self._value * np.ones(self.duration)
+        # print(res)
+        return res
```
Comment (HGSilveri) [inline pulser-core/pulser/waveforms.py:514]: I guess you don't need this anymore

**Review by HGSilveri — COMMENTED** [review]
I think you had some changes that I ended up doing but they got applied anyway. I'll let you decided which ones actually make sense to keep

File: pulser-core/pulser/channels/dmm.py (hunk 1 above)
Comment (vytautas-a) [reply pulser-core/pulser/channels/dmm.py:130]: It gave me a mypy error related to incompatibility of types.

File: pulser-core/pulser/pulse.py (hunk 3 above)
Comment (vytautas-a) [reply pulser-core/pulser/pulse.py:106]: I think phase should be treated on the same footing as amplitude and detuning. From the perspective of pulser-diff, user might do something like `phase=torch.rand(1, requires_grad=True)` and pass this tensor to pulse constructor. However, the phase here will be a 1D tensor and your implementation will fail. I suggest then changing your version to `phase_ := pm.AbstractArray(phase, dtype=float)).size != 1`. This will handle my case.

Comment (vytautas-a) [reply pulser-core/pulser/pulse.py]: Yes, your example has `ndim=0` because you create a tensor from a single float. Mine in the comment above has actually `ndim=1` because that's the way `rand` is implemented. At least in my local version of `torch` :)

File: pulser-core/pulser/math/abstract_array.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def __init__(self, array: AbstractArrayLike, dtype: DTypeLike = None):
         self._array: np.ndarray | torch.Tensor
         if isinstance(array, AbstractArray):
             self._array = array._array
-        elif self.has_torch() and type(array) is torch.Tensor:
+        elif self.has_torch() and isinstance(array, torch.Tensor):
```
Comment (HGSilveri) [inline pulser-core/pulser/math/abstract_array.py:45]: So torch Tensors have subclasses?

File: pulser-core/pulser/math/abstract_array.py [hunk 6]
```diff
@@ -236,6 +241,10 @@ def __setitem__(self, indices: Any, values: AbstractArrayLike) -> None:
         self._array = array
         del self.is_tensor  # Clears cache
 
+    def __iter__(self) -> Generator:
+        for i in self._array:
+            yield i
+
```
Comment (HGSilveri) [inline pulser-core/pulser/math/abstract_array.py:247]: I think that a class with `__getitem__` and `__len__` is automatically iterable. What's the usecase for defining `__iter__` explicitly?

File: pulser-core/pulser/register/_coordinates.py [hunk 7]
```diff
@@ -24,41 +28,44 @@ class CoordsCollection:
         _coords: The coordinates.
     """
 
-    _coords: np.ndarray | list
+    _coords: pm.AbstractArray | list
```
Comment (HGSilveri) [inline pulser-core/pulser/register/_coordinates.py:31]: I would use just `AbstractArrayLike` instead

File: pulser-core/pulser/register/_coordinates.py [hunk 8]
```diff
@@ -24,41 +28,44 @@ class CoordsCollection:
...
 
     @cached_property  # Acts as an attribute in a frozen dataclass
-    def _sorted_coords(self) -> np.ndarray:
-        coords = np.array(self._coords, dtype=float)
-        rounded_coords = np.round(coords, decimals=COORD_PRECISION)
-        sorting = self._calc_sorting_order()
-        return cast(np.ndarray, rounded_coords[sorting])
+    def _sorted_coords(self) -> pm.AbstractArray:
+        coords = pm.vstack(cast(Sequence[AbstractArrayLike], self._coords))
+        rounded_coords = pm.round(coords, decimals=COORD_PRECISION)
```
Comment (HGSilveri) [inline pulser-core/pulser/register/_coordinates.py:47]: Since we are replicating this in `WeightMap`, how about putting these two lines in a `CoordsCollection._rounded_cords` property? You could even cache it

File: pulser-core/pulser/register/weight_maps.py [hunk 9]
```diff
@@ -159,7 +166,8 @@ def _to_abstract_repr(self) -> dict[str, Any]:
             traps=[
                 {"weight": weight, "x": x, "y": y}
                 for weight, (x, y) in zip(
-                    self.sorted_weights, self.sorted_coords
+                    self.sorted_weights,
+                    self.sorted_coords.as_array(detach=True),
```
Comment (HGSilveri) [inline pulser-core/pulser/register/weight_maps.py:170]: I think these sort of changes shouldn't be necessary. I guess if we `detach=True` in `AbstractArray._to_abstract_repr()` it will be fine

**Review by HGSilveri — COMMENTED** [review]
New changes are looking :ok_hand:

File: pulser-core/pulser/math/abstract_array.py (hunk 6 above)
Comment (vytautas-a) [reply pulser-core/pulser/math/abstract_array.py:247]: I got some `mypy` errors related to `AbstractClass` not having `__iter__` . For example, when passing this object to `zip`: `for name, (x, y) in zip(names, self._coords)`.

File: pulser-core/pulser/register/_coordinates.py (hunk 8 above)
Comment (vytautas-a) [reply pulser-core/pulser/register/_coordinates.py:47]: Yes, good idea, I can do that.

File: pulser-core/pulser/register/weight_maps.py (hunk 9 above)
Comment (vytautas-a) [reply pulser-core/pulser/register/weight_maps.py:170]: Correct, the change in this particular place is not necessary - it's a remnant of previous solution. Now `mypy` doesn't complain about it.

File: pulser-core/pulser/math/abstract_array.py (hunk 6 above)
Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:247]: Ok... Well, if we are going to define `__iter__` we have to do it properly. I propose something like this: 
```suggestion
    def __iter__(self) -> Generator[AbstractArray, None, None]]:
        for i in range(self.__len__()):
            yield self.__getitem__(i)

```

File: pulser-core/pulser/math/abstract_array.py (hunk 5 above)
Comment (vytautas-a) [reply pulser-core/pulser/math/abstract_array.py:45]: What do you mean? I made this change since I think it's more pythonic to check for class with `isinstance` - at least in `qadence` we do that. :)

Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:45]: It was a guess of your motivation to do the change, I'm fine with `isinstance` :)

File: pulser-core/pulser/register/_coordinates.py (hunk 7 above)
Comment (vytautas-a) [reply pulser-core/pulser/register/_coordinates.py:31]: If I make such change - a lot of `mypy` errors pop up saying that some types of `AbstractArrayLike` have no needed attributes, e.g., `Item "int" of "Union[Union[Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]], AbstractArray], List[Any]]" has no attribute "__iter__" (not iterable)  [union-attr]`.

I think it's easier to leave `pm.AbstractArray` here.

File: pulser-core/pulser/math/abstract_array.py (hunk 6 above)
Comment (vytautas-a) [reply pulser-core/pulser/math/abstract_array.py:247]: If I include the specification `[AbstractArray, None, None]` for `Generator` type I get these `mypy` errors:
```
pulser-core/pulser/register/base_register.py:91: error: Argument 2 to "zip" has incompatible type "Union[AbstractArray, List[Any]]"; expected "Iterable[ndarray[Any, Any]]"  [arg-type]
pulser-core/pulser/register/base_register.py:261: error: Argument 1 to "tolist" of "_ArrayOrScalarCommon" has incompatible type "Union[AbstractArray, Any]"; expected "_ArrayOrScalarCommon"  [arg-type]
```

Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:247]: These errors look sensible to me, the type hints in `BaseRegister` need to be updated or the methods changed

Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:247]: For example, I would change the `qubits` property to:

```python
    @property
    def qubits(self) -> dict[QubitId, pm.AbstractArray]:
        """Dictionary of the qubit names and their position coordinates."""
        return dict(zip(self._ids, pm.vstack(self._coords)))
```

File: pulser-core/pulser/math/abstract_array.py [hunk 10]
```diff
@@ -89,6 +89,10 @@ def as_array(self, detach: bool = False) -> np.ndarray:
             )
         return np.asarray(self._array)
 
+    def as_list(self) -> list:
```
Comment (HGSilveri) [inline pulser-core/pulser/math/abstract_array.py:92]: We should probably call it `tolist()` instead

File: pulser-core/pulser/register/_coordinates.py [hunk 11]
```diff
@@ -41,12 +41,15 @@ def sorted_coords(self) -> pm.AbstractArray:
         # Copies to prevent direct access to self._sorted_coords
         return self._sorted_coords.copy()
 
+    @cached_property
+    def rounded_coords(self) -> pm.AbstractArray:
```
Comment (HGSilveri) [inline pulser-core/pulser/register/_coordinates.py:45]: I would prefer if this was a private property, unless you thought of a case where it makes sense to have it public

File: pulser-core/pulser/register/_coordinates.py [hunk 12]
```diff
@@ -41,12 +41,15 @@ def sorted_coords(self) -> pm.AbstractArray:
...
+        coords = pm.vstack(cast(Sequence[AbstractArrayLike], self._coords))
+        return pm.round(coords, decimals=COORD_PRECISION)
+
     @cached_property  # Acts as an attribute in a frozen dataclass
     def _sorted_coords(self) -> pm.AbstractArray:
-        coords = pm.vstack(cast(Sequence[AbstractArrayLike], self._coords))
-        rounded_coords = pm.round(coords, decimals=COORD_PRECISION)
-        sorting = self._calc_sorting_order(rounded_coords)
-        return rounded_coords[sorting]
+        sorting = self._calc_sorting_order(self.rounded_coords)
```
Comment (HGSilveri) [inline pulser-core/pulser/register/_coordinates.py:51]: I guess `_calc_sorting_order()` no longer needs to receive `rounded_coords`, right?

File: pulser-core/pulser/register/base_register.py [hunk 13]
```diff
@@ -258,7 +265,7 @@ def _to_dict(self) -> dict[str, Any]:
         return obj_to_dict(
             self,
             cls_dict,
-            [np.ndarray.tolist(qubit_coords) for qubit_coords in self._coords],
+            [qubit_coords.as_list() for qubit_coords in self._coords],
```
Comment (HGSilveri) [inline pulser-core/pulser/register/base_register.py:268]: Does this fail for tensors that require grad?

File: pulser-simulation/pulser_simulation/hamiltonian.py [hunk 14]
```diff
@@ -46,14 +47,14 @@ class Hamiltonian:
...
-        qdict: dict[QubitId, np.ndarray],
+        qdict: dict[QubitId, pm.AbstractArray],
         device: BaseDevice,
         sampling_rate: float,
         config: NoiseModel,
     ) -> None:
         """Instantiates a Hamiltonian object."""
         self.samples_obj = samples_obj
-        self._qdict = qdict
+        self._qdict = {k: v.as_array() for k, v in qdict.items()}
```
Comment (HGSilveri) [inline pulser-simulation/pulser_simulation/hamiltonian.py:57]: Should you detach here?

File: pulser-core/pulser/register/base_register.py (hunk 13 above)
Comment (vytautas-a) [reply pulser-core/pulser/register/base_register.py:268]: I changed this line not due to functionality issues but `mypy` error.

File: pulser-simulation/pulser_simulation/hamiltonian.py (hunk 14 above)
Comment (vytautas-a) [reply pulser-simulation/pulser_simulation/hamiltonian.py:57]: correct

File: pulser-core/pulser/register/_coordinates.py (hunk 12 above)
Comment (vytautas-a) [reply pulser-core/pulser/register/_coordinates.py:51]: correct

File: pulser-core/pulser/register/base_register.py (hunk 13 above)
Comment (HGSilveri) [reply pulser-core/pulser/register/base_register.py:268]: I get that, but I was just wondering if `tolist()` failed for tensors that required grad. I just checked that it just doesn't care

File: pulser-core/pulser/parametrized/variable.py [hunk 15]
```diff
@@ -71,9 +73,10 @@ def _assign(self, value: Union[ArrayLike, float, int]) -> None:
         object.__setattr__(self, "_count", self._count + 1)
 
     def _validate_value(
-        self, value: Union[ArrayLike, float, int]
-    ) -> np.ndarray:
-        val = np.array(value, dtype=self.dtype, ndmin=1)
+        self, value: Union[AbstractArrayLike, float, int]
+    ) -> pm.AbstractArray:
+        val = pm.AbstractArray(value)
+        # val = np.array(value, dtype=self.dtype, ndmin=1)
```
Comment (HGSilveri) [inline pulser-core/pulser/parametrized/variable.py:79]: The casting to `self.dtype` and the `ndmin=1` are both important. `AbstractArray` already has a `dtype` argument, we should add `ndmin` too

File: pulser-core/pulser/register/_coordinates.py [hunk 16]
```diff
@@ -24,41 +28,45 @@ class CoordsCollection:
...
     def _calc_sorting_order(self) -> np.ndarray:
         """Calculates the unique order that sorts the coordinates."""
-        coords = np.array(self._coords, dtype=float)
         # Sorting the coordinates 1st left to right, 2nd bottom to top
-        rounded_coords = np.round(coords, decimals=COORD_PRECISION)
-        dims = rounded_coords.shape[1]
-        sorter = [rounded_coords[:, i] for i in range(dims - 1, -1, -1)]
+        dims = self._rounded_coords.shape[1]
+        sorter = [
+            self._rounded_coords.as_array(detach=True)[:, i]
```
Comment (HGSilveri) [inline pulser-core/pulser/register/_coordinates.py:59]: This is detaching for every `i`, it would be better to detach only once and store the array in a variable

**Review by HGSilveri — COMMENTED** [review]
Last comment and we can merge, great job!

File: pulser-core/pulser/parametrized/variable.py (hunk 15 above)
Comment (vytautas-a) [reply pulser-core/pulser/parametrized/variable.py:79]: Ok, I can do that. But what's the actual reason for this?

Comment (HGSilveri) [reply pulser-core/pulser/parametrized/variable.py:79]: A variable's value is always a 1D array, even when it has only one entry. We accept 0D values but then internally turn them into a single-entry 1D array through this `ndmin=1` option

Comment (vytautas-a) [reply pulser-core/pulser/parametrized/variable.py:79]: yes, makes sense

Comment (vytautas-a) [reply pulser-core/pulser/parametrized/variable.py:79]: What if we skip this argument completely and simply make sure that any tensor/array gets reshaped to 1D shape if it is 0D? I noticed that `ndmin` is used only in `variable.py` and with value 1, so I think there's no need to fully replicate the behavior of `ndmin` of numpy here.

Comment (HGSilveri) [reply pulser-core/pulser/parametrized/variable.py:79]: > What if we skip this argument completely and simply make sure that any tensor/array gets reshaped to 1D shape if it is 0D?

I think that is a strong and potentially problematic assumption that I would prefer to avoid making in general.

> I noticed that `ndmin` is used only in `variable.py` and with value 1, so I think there's no need to fully replicate the behavior of `ndmin` of numpy here.

I agree that there's no need to fully replicate the behaviour of `ndmin`. You can make a restricted `force_array: bool` argument for this purpose, how about that?

File: pulser-core/pulser/math/abstract_array.py [hunk 17]
```diff
@@ -37,19 +37,27 @@ class AbstractArray:
...
+        elif self.has_torch() and isinstance(array, torch.Tensor):
             self._array = torch.as_tensor(
                 array,
                 dtype=dtype,  # type: ignore[arg-type]
             )
         else:
             self._array = np.asarray(array, dtype=dtype)
 
+        if force_array:
+            self._array = self._array[None]
```
Comment (HGSilveri) [inline pulser-core/pulser/math/abstract_array.py:59]: We only want to do this for scalars:
```suggestion
        if force_array and self._array.ndim == 0:
            self._array = self._array[None]
```

**Review by HGSilveri — APPROVED** [review]
Perfect, let's merge! Thanks a lot for your efforts
